File: rrd_widgets/components/widget/Button/SwitchButton.py
from PySide6.QtGui import QColor, QPainter, QPen
from PySide6.QtWidgets import QToolButton, QWidget
from PySide6.QtCore import Qt, Property, QPoint, QPropertyAnimation, QRect, Signal


class SwitchButtonBase(QToolButton):
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)
        self.setCheckable(True)
        self._indicatorX = 0

        self.animaInit()

    def animaInit(self):
        self.slideAni = QPropertyAnimation(self, b'indicatorX')
        self.slideAni.setDuration(200)

    def animaParamsInit(self):
        self.start = self.height() // 2
        self.end = self.width() - 2 - self.start
        self._indicatorX = self.start

    def setParams(self):
        pass

    def toggle(self):
        self.setChecked(not self.isChecked())
        if self.isChecked():
            self.slideAni.setEndValue(self.end)
        else:
            self.slideAni.setEndValue(self.start)
        self.slideAni.start()

    def mouseReleaseEvent(self, event) -> None:
        super().mouseReleaseEvent(event)
        if self.isChecked():
            self.slideAni.setEndValue(self.end)
        else:
            self.slideAni.setEndValue(self.start)
        self.slideAni.start()

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHints(QPainter.Antialiasing)
        self.paintBackground(painter)
        self.paintIndicator(painter)

    def paintBackground(self, painter: QPainter):
        pass

    def paintIndicator(self, painter: QPainter):
        pass

    @Property(float)
    def indicatorX(self):
        return self._indicatorX

    @indicatorX.setter
    def indicatorX(self, x):
        self._indicatorX = x
        self.update()

    def resizeEvent(self, event) -> None:
        super().resizeEvent(event)
        self.animaParamsInit()


class SwitchButton_1(SwitchButtonBase):
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)

        self.indicator_color = QColor()
        self.background_color = QColor()
        self.checked_background_color = QColor()

    def setParams(self,
                  indicator_color: QColor = QColor(255, 255, 255),
                  background_color: QColor = QColor(188, 188, 188),
                  checked_background_color: QColor = QColor(0, 89, 89),
                  checked_indicator_color: QColor = QColor(255,255,255)
                  ):
        self.indicator_color = indicator_color
        self.background_color = background_color
        self.checked_background_color = checked_background_color
        self.checked_indicator_color = checked_indicator_color

    def paintBackground(self, painter: QPainter):
        painter.save()
        r = self.height() // 2
        rect = QRect(1, 1, self.width() - 2, self.height() - 1)

        if not self.isChecked():
            painter.setBrush(self.background_color)
        else:
            painter.setBrush(self.checked_background_color)

        painter.setPen(Qt.NoPen)
        painter.drawRoundedRect(rect, r, r)
        painter.restore()

    def paintIndicator(self, painter: QPainter):
        painter.save()
        r = self.height() // 3
        point = QPoint(int(self._indicatorX), 0)

        painter.translate(0, self.height() // 2)
        painter.setPen(Qt.NoPen)

        if self.isChecked():
            painter.setBrush(self.checked_indicator_color)
        else:
            painter.setBrush(self.indicator_color)

        painter.drawEllipse(point, r, r)
        painter.restore()


class SwitchButton_2(SwitchButtonBase):
    def setParams(self,
                  indicator_color: QColor = QColor(181, 181, 181),
                  background_color: QColor = QColor(181, 181, 181),
                  checked_background_color: QColor = QColor(255, 167, 38),
                  checked_indicator_color: QColor = QColor(255, 167, 38)
                  ):
        self.indicator_color = indicator_color
        self.background_color = background_color
        self.checked_background_color = checked_background_color
        self.checked_indicator_color = checked_indicator_color

    def paintBackground(self, painter: QPainter):
        painter.save()
        r = self.height() // 2
        rect = QRect(1, 1, self.width() - 2, self.height() - 2)

        pen = QPen()
        pen.setWidth(3)
        if self.isChecked():
            pen.setColor(self.checked_background_color)
            painter.setPen(pen)
        else:
            pen.setColor(self.background_color)
            painter.setPen(pen)

        painter.drawRoundedRect(rect, r, r)
        painter.restore()

    def paintIndicator(self, painter: QPainter):
        painter.save()
        r = self.height() // 3
        point = QPoint(int(self._indicatorX), 0)

        painter.translate(0, self.height() // 2)

        if self.isChecked():
            painter.setBrush(self.checked_indicator_color)
        else:
            painter.setBrush(self.indicator_color)
        painter.setPen(Qt.NoPen)
        painter.drawEllipse(point, r, r)
        painter.restore()


class SwitchButton_3(SwitchButtonBase):
    def setParams(self,
                  indicator_color: QColor = QColor(255, 255, 255),
                  background_color: QColor = QColor(204, 204, 204),
                  checked_background_color: QColor = QColor(51, 102, 153, 200),
                  checked_indicator_color: QColor = QColor(51, 102, 153)
                  ):
        self.indicator_color = indicator_color
        self.background_color = background_color
        self.checked_background_color = checked_background_color
        self.checked_indicator_color = checked_indicator_color

    def paintBackground(self, painter: QPainter):
        painter.save()

        w = self.width() - self.width() // 3
        h = self.height() - self.height() // 3
        rect = QRect(-w // 2, -h // 2, w, h)
        r = h // 2

        painter.translate(self.width() // 2, self.height() // 2)

        if self.isChecked():
            painter.setBrush(self.checked_background_color)
        else:
            painter.setBrush(self.background_color)
        painter.setPen(Qt.NoPen)
        painter.drawRoundedRect(rect, r, r)
        painter.restore()

    def paintIndicator(self, painter: QPainter):
        painter.save()
        r = (self.height() - self.height() // 3) // 2 + self.height() // 9
        point = QPoint(int(self._indicatorX), 0)

        painter.translate(0, self.height() // 2)

        if self.isChecked():
            painter.setBrush(self.checked_indicator_color)
        else:
            painter.setBrush(self.indicator_color)
        painter.setPen(Qt.NoPen)
        painter.drawEllipse(point, r, r)
        painter.restore()


class SwitchButton_4(SwitchButtonBase):
    def animaParamsInit(self):
        super().animaParamsInit()
        self.isHover = False

    def setParams(self,
                  indicator_color: QColor = QColor(255, 255, 255),
                  background_color: QColor = QColor(204, 204, 204),
                  checked_background_color: QColor = QColor(153, 0, 51, 200),
                  checked_indicator_color: QColor = QColor(153, 0, 51)
                  ):
        self.indicator_color = indicator_color
        self.background_color = background_color
        self.checked_background_color = checked_background_color
        self.checked_indicator_color = checked_indicator_color

    def enterEvent(self, event):
        super().enterEvent(event)
        self.isHover = True

    def leaveEvent(self, event):
        super().leaveEvent(event)
        self.isHover = False

    def paintBackground(self, painter: QPainter):
        painter.save()

        w = self.width() - self.width() // 2
        h = self.height() - self.height() // 2

        r = h // 2

        painter.translate(self.width() // 2, self.height() // 2)

        rect = QRect(-w // 2, -h // 2, w, h)

        painter.setPen(Qt.NoPen)
        if self.isChecked():
            painter.setBrush(self.checked_background_color)
        else:
            painter.setBrush(self.background_color)

        painter.drawRoundedRect(rect, r, r)
        painter.restore()

    def paintIndicator(self, painter: QPainter):
        painter.save()
        r = (self.height() - self.height() // 2) // 2 + self.height() // 9
        point = QPoint(int(self._indicatorX), 0)

        painter.translate(0, self.height() // 2)
        painter.setPen(Qt.NoPen)
        if self.isChecked():
            if self.isHover:
                c = self.checked_indicator_color
                painter.setBrush(QColor(c.red(), c.green(), c.blue(), 100))
                painter.drawEllipse(point, r + 5, r + 5)
            painter.setBrush(self.checked_indicator_color)

        else:
            if self.isHover:
                c = self.indicator_color
                painter.setBrush(QColor(c.red(), c.green(), c.blue(), 100))
                painter.drawEllipse(point, r + 5, r + 5)
            painter.setBrush(self.indicator_color)

        painter.drawEllipse(point, r, r)

        painter.restore()


class SwitchButton_5(SwitchButtonBase):
    def setParams(self,
                  indicator_color: QColor = QColor(255, 255, 255),
                  background_color: QColor = QColor(204, 204, 204),
                  checked_background_color: QColor = QColor(153, 0, 51, 200),
                  ):
        self.indicator_color = indicator_color
        self.background_color = background_color
        self.checked_background_color = checked_background_color

    def animaParamsInit(self):
        self.var = self.height() // 8
        h = self.height() - self.var * 2
        self.start = self.var
        self.end = self.width() - h - self.var
        self._indicatorX = self.var

    def paintBackground(self, painter: QPainter):
        painter.save()
        if self.isChecked():
            painter.setBrush(self.checked_background_color)
        else:
            painter.setBrush(self.background_color)

        painter.setPen(Qt.NoPen)
        painter.drawRoundedRect(self.rect(), 5, 5)
        painter.restore()

    def paintIndicator(self, painter: QPainter):
        painter.save()
        h = self.height() - self.var * 2
        rect = QRect(self._indicatorX, -h // 2, h, h)

        painter.translate(0, self.height() // 2)
        painter.setPen(Qt.NoPen)
        painter.setBrush(self.indicator_color)

        painter.drawRoundedRect(rect, 5, 5)
        painter.restore()

File: c++/rrd_widgets/components/widget/Button/src/SwitchButton.cpp
/*
 * Qt-RoastedDuck - Widgets
 * ======================
 * Qt widgets - based implementation of the Material Design specification.
 * 
 * Repository at https ://github.com/Rev-RoastedDuck/Qt-RoastedDuck-Widgets.
 * 
 * Demo are available at https ://github.com/Rev-RoastedDuck/Qt-RoastedDuck-Widgets/tree/main/Demo.
 * 
 * Examples are available at https ://github.com/Rev-RoastedDuck/Qt-RoastedDuck-Widgets/tree/main/examples.
 * 
 * Information:
 * WeChat: Roast_71.
 * csdnBlog : https ://blog.csdn.net/m0_72760466?type=blog.
 * 
 * 	: copyright : (c)2023 by Rev - RoastedDuck.
 * 	: license : GPLv3, see LICENSE for more details.
*/

#include "SwitchButton.h"

/******************************************************************************/
/*---------------------------- SwitchButton Base -----------------------------*/
/******************************************************************************/
SwitchButtonBase::SwitchButtonBase(QWidget *parent)
    : QToolButton(parent), _indicatorX(0.0f)
{
    setCheckable(true);
    animaInit();
}

void SwitchButtonBase::mouseReleaseEvent(QMouseEvent *event)
{
    QToolButton::mouseReleaseEvent(event);
    if (isChecked())
        slideAni->setEndValue(end);
    else
        slideAni->setEndValue(start);
    slideAni->start();
}

void SwitchButtonBase::paintEvent(QPaintEvent *event)
{
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    paintBackground(painter);
    paintIndicator(painter);
}

void SwitchButtonBase::resizeEvent(QResizeEvent *event)
{
    QToolButton::resizeEvent(event);
    animaParamsInit();
}

float SwitchButtonBase::indicatorX() const
{
    return _indicatorX;
}

void SwitchButtonBase::setIndicatorX(float x)
{
    _indicatorX = x;
    update();
    emit indicatorXChanged();
}

void SwitchButtonBase::animaInit()
{
    slideAni = new QPropertyAnimation(this, "indicatorX");
    slideAni->setDuration(200);
}

void SwitchButtonBase::animaParamsInit()
{
    start = height() / 2;
    end = width() - 2 - start;
    _indicatorX = start;
}


/******************************************************************************/
/*------------------------------ SwitchButton --------------------------------*/
/******************************************************************************/
/** \addtogroup SwitchButton_1
 * \{ */
SwitchButton_1::SwitchButton_1(QWidget *parent)
    : SwitchButtonBase(parent)
{
    indicator_color = QColor(255, 255, 255);
    background_color = QColor(188, 188, 188);
    checked_background_color = QColor(0, 89, 89);
    checked_indicator_color = QColor(255, 255, 255);
}

void SwitchButton_1::setParams(const QColor &indicatorColor,
                               const QColor &backgroundColor,
                               const QColor &checkedBackgroundColor,
                               const QColor &checkedIndicatorColor)
{
    indicator_color = indicatorColor;
    background_color = backgroundColor;
    checked_background_color = checkedBackgroundColor;
    checked_indicator_color = checkedIndicatorColor;
}

void SwitchButton_1::paintBackground(QPainter &painter)
{
    painter.save();
    int r = height() / 2;
    QRect rect(1, 1, width() - 2, height() - 1);

    if (!isChecked())
        painter.setBrush(background_color);
    else
        painter.setBrush(checked_background_color);

    painter.setPen(Qt::NoPen);
    painter.drawRoundedRect(rect, r, r);
    painter.restore();
}

void SwitchButton_1::paintIndicator(QPainter &painter)
{
    painter.save();
    int r = height() / 3;
    QPoint point(static_cast<int>(_indicatorX), 0);

    painter.translate(0, height() / 2);
    painter.setPen(Qt::NoPen);

    if (isChecked())
        painter.setBrush(checked_indicator_color);
    else
        painter.setBrush(indicator_color);

    painter.drawEllipse(point, r, r);
    painter.restore();
}
/** \} */


/** \addtogroup SwitchButton_2
 * \{ */
SwitchButton_2::SwitchButton_2(QWidget *parent)
    : SwitchButtonBase(parent)
{
    indicator_color = QColor(181, 181, 181);
    background_color = QColor(181, 181, 181);
    checked_background_color = QColor(0, 89, 89);
    checked_indicator_color = QColor(0, 89, 89);
}

void SwitchButton_2::setParams(const QColor &indicatorColor,
                               const QColor &backgroundColor,
                               const QColor &checkedBackgroundColor,
                               const QColor &checkedIndicatorColor)
{
    indicator_color = indicatorColor;
    background_color = backgroundColor;
    checked_background_color = checkedBackgroundColor;
    checked_indicator_color = checkedIndicatorColor;
}

void SwitchButton_2::paintBackground(QPainter &painter)
{
    painter.save();
    int r = height() / 2;
    QRect rect(1, 1, width() - 2, height() - 2);

    QPen pen;
    pen.setWidth(3);
    if (isChecked())
    {
        pen.setColor(checked_background_color);
        painter.setPen(pen);
    }
    else
    {
        pen.setColor(background_color);
        painter.setPen(pen);
    }

    painter.drawRoundedRect(rect, r, r);
    painter.restore();
}

void SwitchButton_2::paintIndicator(QPainter &painter)
{
    painter.save();
    int r = height() / 3;
    QPoint point(static_cast<int>(_indicatorX), 0);

    painter.translate(0, height() / 2);

    if (isChecked())
        painter.setBrush(checked_indicator_color);
    else
        painter.setBrush(indicator_color);

    painter.setPen(Qt::NoPen);
    painter.drawEllipse(point, r, r);
    painter.restore();
}

/** \} */



/** \addtogroup SwitchButton_3
 * \{ */
SwitchButton_3::SwitchButton_3(QWidget *parent)
    : SwitchButtonBase(parent)
{
    indicator_color = QColor(255, 255, 255);
    background_color = QColor(204, 204, 204);
    checked_background_color = QColor(0, 89, 89);
    checked_indicator_color = QColor(255, 255, 255);
}

void SwitchButton_3::setParams(const QColor &indicatorColor,
                               const QColor &backgroundColor,
                               const QColor &checkedBackgroundColor,
                               const QColor &checkedIndicatorColor)
{
    indicator_color = indicatorColor;
    background_color = backgroundColor;
    checked_background_color = checkedBackgroundColor;
    checked_indicator_color = checkedIndicatorColor;
}

void SwitchButton_3::paintBackground(QPainter &painter)
{
    painter.save();

    int w = width() - width() / 3;
    int h = height() - height() / 3;
    QRect rect(-w / 2, -h / 2, w, h);
    int r = h / 2;

    painter.translate(width() / 2, height() / 2);

    if (isChecked())
        painter.setBrush(checked_background_color);
    else
        painter.setBrush(background_color);

    painter.setPen(Qt::NoPen);
    painter.drawRoundedRect(rect, r, r);
    painter.restore();
}

void SwitchButton_3::paintIndicator(QPainter &painter)
{
    painter.save();
    int r = (height() - height() / 3) / 2 + height() / 9;
    QPoint point(static_cast<int>(_indicatorX), 0);

    painter.translate(0, height() / 2);

    if (isChecked())
        painter.setBrush(checked_indicator_color);
    else
        painter.setBrush(indicator_color);

    painter.setPen(Qt::NoPen);
    painter.drawEllipse(point, r, r);
    painter.restore();
}


/** \} */

/** \addtogroup SwitchButton_4
 * \{ */
SwitchButton_4::SwitchButton_4(QWidget *parent)
    : SwitchButtonBase(parent), isHover(false)
{
    indicator_color = QColor(255, 255, 255);
    background_color = QColor(204, 204, 204);
    checked_background_color = QColor(153, 0, 51, 200);
    checked_indicator_color = QColor(255, 255, 255);
}

void SwitchButton_4::setParams(const QColor &indicatorColor,
                               const QColor &backgroundColor,
                               const QColor &checkedBackgroundColor,
                               const QColor &checkedIndicatorColor)
{
    indicator_color = indicatorColor;
    background_color = backgroundColor;
    checked_background_color = checkedBackgroundColor;
    checked_indicator_color = checkedIndicatorColor;
}

void SwitchButton_4::paintBackground(QPainter &painter)
{
    painter.save();

    int w = width() - width() / 2;
    int h = height() - height() / 2;
    int r = h / 2;

    painter.translate(width() / 2, height() / 2);

    QRect rect(-w / 2, -h / 2, w, h);

    painter.setPen(Qt::NoPen);
    if (isChecked())
        painter.setBrush(checked_background_color);
    else
        painter.setBrush(background_color);

    painter.drawRoundedRect(rect, r, r);
    painter.restore();
}

void SwitchButton_4::paintIndicator(QPainter &painter)
{
    painter.save();
    int r = (height() - height() / 2) / 2 + height() / 9;
    QPoint point(static_cast<int>(_indicatorX), 0);

    painter.translate(0, height() / 2);
    painter.setPen(Qt::NoPen);

    if (isChecked())
    {
        if (isHover)
        {
            QColor c = checked_indicator_color;
            painter.setBrush(QColor(c.red(), c.green(), c.blue(), 100));
            painter.drawEllipse(point, r + 5, r + 5);
        }
        painter.setBrush(checked_indicator_color);
    }
    else
    {
        if (isHover)
        {
            QColor c = indicator_color;
            painter.setBrush(QColor(c.red(), c.green(), c.blue(), 100));
            painter.drawEllipse(point, r + 5, r + 5);
        }
        painter.setBrush(indicator_color);
    }

    painter.drawEllipse(point, r, r);
    painter.restore();
}

void SwitchButton_4::enterEvent(QEnterEvent *event)
{
    SwitchButtonBase::enterEvent(event);
    isHover = true;
}

void SwitchButton_4::leaveEvent(QEvent *event)
{
    SwitchButtonBase::leaveEvent(event);
    isHover = false;
}

void SwitchButton_4::animaParamsInit()
{
    SwitchButtonBase::animaParamsInit();
    isHover = false;
}

/** \} */



/** \addtogroup SwitchButton_5
 * \{ */
SwitchButton_5::SwitchButton_5(QWidget *parent)
    : SwitchButtonBase(parent)
{
    indicator_color = QColor(255, 255, 255);
    background_color = QColor(204, 204, 204);
    checked_background_color = QColor(153, 0, 51, 200);
}

void SwitchButton_5::setParams(const QColor &indicatorColor,
                               const QColor &backgroundColor,
                               const QColor &checkedBackgroundColor)
{
    indicator_color = indicatorColor;
    background_color = backgroundColor;
    checked_background_color = checkedBackgroundColor;
}

void SwitchButton_5::animaParamsInit()
{
    var = height() / 8;
    int h = height() - var * 2;
    start = var;
    end = width() - h - var;
    _indicatorX = var;
}

void SwitchButton_5::paintBackground(QPainter &painter)
{
    painter.save();

    if (isChecked())
        painter.setBrush(checked_background_color);
    else
        painter.setBrush(background_color);

    painter.setPen(Qt::NoPen);
    painter.drawRoundedRect(rect(), 5, 5);
    painter.restore();
}

void SwitchButton_5::paintIndicator(QPainter &painter)
{
    painter.save();
    int h = height() - var * 2;
    QRect rect(static_cast<int>(_indicatorX), -h / 2, h, h);

    painter.translate(0, height() / 2);
    painter.setPen(Qt::NoPen);
    painter.setBrush(indicator_color);

    painter.drawRoundedRect(rect, 5, 5);
    painter.restore();
}
/** \} */